src/util/apparmor.c

```diff
@@ -93,3 +105,453 @@ int bus_apparmor_dbus_supported(bool *supportedp) {
+                                                                interface, method, allow, &audit_tmp);
+                                if (r < 0)
+                                        return r;
+                                if (!allow)
+                                        return r;
+                                if (audit_tmp)
+                                        *audit = 1;
+                        }
+                }
+        } else if (nameset->type == NAME_SET_TYPE_SNAPSHOT) {
+                /*
+                 * An ownership-snapshot is given. Again, we simply iterate the
+                 * names and match each. Note that the snapshot must contain
+                 * queued names as well, since the policy matches on it.
+                 */
+                for (i = 0; i < nameset->snapshot->n_names; ++i) {
+                        r = apparmor_message_query_name(check_send, security_label, bustype,
+                                                        receiver_context,
+                                                        nameset->snapshot->names[i]->name,
+                                                        path, interface, method, allow, &audit_tmp);
+                        if (r < 0)
+                                return r;
+                        if (!allow)
+                                return r;
+                        if (audit_tmp)
+                                *audit = 1;
+                }
+        } else if (nameset->type != NAME_SET_TYPE_EMPTY) {
+                c_assert(0);
+                r = -EINVAL;
+        }
+
+        return r;
+}
+
+/**
+ * bus_apparmor_check_own() - check if the given transaction is allowed
+ * @registry:           AppArmor registry to operate on
+ * @context:            security context requesting the name
+ * @name:               name to be owned
+ *
+ * Check if the given owner context is allowed to own the given name.
+ *
+ * The contexts are pinned when peers connect, and as such could in principle
+ * become invalid in case a new policy is loaded that does not know the
+ * old labels. In this case we treat this as if the ownership request was
+ * denied.
+ *
+ * Return: 0 if the ownership is allowed, APPARMOR_E_DENIED if it is not,
+ *         or a negative error code on failure.
+ */
+int bus_apparmor_check_own(struct BusAppArmorRegistry *registry,
+                           const char *owner_context,
+                           const char *name) {
+        _c_cleanup_(c_freep) char *condup = strdup(owner_context);
+        char *security_label, *security_mode;
+        _c_cleanup_(c_freep) char *qstr = NULL;
+        int r;
+        /* the AppArmor API uses pointers to int for pointers to boolean */
+        int allow = false, audit = true;
+
+        if (!is_apparmor_enabled())
+                return 0;
+
+        if (!condup)
+                return -ENOMEM;
```

Comment:
`error_origin(-ENOMEM)`, and I would suggest moving the `strdup()` out of the variable declaration.